zipkin-server/src/main/java/zipkin2/server/internal/throttle/ThrottledCall.java

```diff
@@ -0,0 +1,240 @@
+ * purposes:
+ * <ol>
+ * <li>Limits the number of requests that can run in parallel.</li>
+ * <li>Depending on configuration, can queue up requests to make sure we don't aggressively drop requests that would
+ *     otherwise succeed if given a moment. Bounded queues are safest for this as unbounded ones can lead to heap
+ *     exhaustion and {@link OutOfMemoryError OOM errors}.</li>
+ * </ol>
+ *
+ * @see ThrottledStorageComponent
+ */
+final class ThrottledCall<V> extends Call<V> {
+  final ExecutorService executor;
+  final Limiter<Void> limiter;
+  final Listener limitListener;
+  /**
+   * Delegate call needs to be supplied later to avoid having it take action when it is created (like
+   * {@link InMemoryStorage} and thus avoid being throttled.
+   */
+  final Supplier<Call<V>> delegate;
+  Call<V> call;
+  volatile boolean canceled;
+
+  public ThrottledCall(ExecutorService executor, Limiter<Void> limiter, Supplier<Call<V>> delegate) {
+    this.executor = executor;
+    this.limiter = limiter;
+    this.limitListener = limiter.acquire(null).orElseThrow(RejectedExecutionException::new);
+    this.delegate = delegate;
+  }
+
+  private ThrottledCall(ThrottledCall other) {
+    this(other.executor, other.limiter, other.call == null ? other.delegate : () -> other.call.clone());
+  }
+
+  @Override
+  public V execute() throws IOException {
+    try {
+      call = delegate.get();
+
+      // Make sure we throttle
+      Future<V> future = executor.submit(() -> {
+        String oldName = setCurrentThreadName(call.toString());
+        try {
+          return call.execute();
+        } finally {
+          setCurrentThreadName(oldName);
+        }
+      });
+      V result = future.get(); // Still block for the response
+
+      limitListener.onSuccess();
+      return result;
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof RejectedExecutionException) {
+        // Storage rejected us, throttle back
+        limitListener.onDropped();
+      } else {
+        limitListener.onIgnore();
+      }
+
+      if (cause instanceof RuntimeException) {
+        throw (RuntimeException) cause;
+      } else if (cause instanceof IOException) {
+        throw (IOException) cause;
+      } else {
+        throw new RuntimeException("Issue while executing on a throttled call", cause);
+      }
+    } catch (InterruptedException e) {
+      limitListener.onIgnore();
+      throw new RuntimeException("Interrupted while blocking on a throttled call", e);
+    } catch (Exception e) {
+      // Ignoring in all cases here because storage itself isn't saying we need to throttle.  Though, we may still be
+      // write bound, but a drop in concurrency won't necessarily help.
+      limitListener.onIgnore();
+      throw e;
+    }
+  }
+
+  @Override
+  public void enqueue(Callback<V> callback) {
+    try {
+      executor.execute(new QueuedCall(callback));
+    } catch (Exception e) {
```

Comment:
similar, please don't catch Exception unless checked exceptions are declared as it is distracting. Probably you want `RuntimeException|Error` with `Call.propagateIfFatal` guard